marimo/_save/toplevel.py

```diff
@@ -0,0 +1,48 @@
+# Copyright 2025 Marimo. All rights reserved.
+from __future__ import annotations
+
+from typing import Any, Callable
+
+from marimo import _loggers
+from marimo._ast.app import App, InternalApp
+from marimo._ast.compiler import toplevel_cell_factory
+from marimo._ast.names import (
+    SELF_CELL_NAME,
+    SETUP_CELL_NAME,
+    TOPLEVEL_CELL_PREFIX,
+)
+from marimo._ast.toplevel import TopLevelExtraction
+from marimo._runtime.dataflow import DirectedGraph
+from marimo._types.ids import CellId_t
+
+LOGGER = _loggers.marimo_logger()
+
+
+def graph_from_app(fn: Callable[..., Any], app: App) -> DirectedGraph:
```

Comment:
I wouldn't mind a bit of a sanity check here.

This is the setup: we are loading a marimo notebook as a module
a toplevel cache decorated function has dependencies on other top level dependencies.

This function is called on first invocation, where we do a best effort to create the graph thus far with top level definitions.

Notably, function definitions can be out of order (solved by waiting for invocation), but also invoked even before the app is done being constructed (via decorators)